src/libstore/misc.cc

```diff
@@ -200,6 +200,40 @@ void Store::queryMissing(const std::vector<DerivedPath> & targets,
             auto drv = make_ref<Derivation>(derivationFromPath(bfd.drvPath));
             ParsedDerivation parsedDrv(StorePath(bfd.drvPath), *drv);
 
+            if (!knownOutputPaths && settings.useSubstitutes && parsedDrv.substitutesAllowed()) {
+                // Try to realise the static output hashes.
+
+                // TODO: I'm not sure this logic is quite right, how do we
+                // ensure all the needed paths are present?
```

Comment:
I'm happier with the current logic though still not super confident. My reasoning/intent is that if we find realisations for all of the requested output paths, then we can fall back to the `checkOutput` code; but maybe I should be checking for isValidPath as is done above. I'm not sure how 'valid' is defined here, it just means that it's present in the store already, I think?

Reply:
Strictly speaking, we should check whether the path is valid (as we don't need to re-fetch it in that case).
As a first approximation, I think we can just pretend that the derivation output is valid if the output path is.